autofill/autofill-api/src/main/java/com/duckduckgo/autofill/api/AutofillCapabilityChecker.kt

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2023 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.autofill.api
+
+/**
+ * Used to check the status of various Autofill features.
+ *
+ * Whether autofill features are enabled depends on a variety of inputs. This class provides a single way to query the status of all of them.
+ */
+interface AutofillCapabilityChecker {
+
+    /**
+     * Whether autofill is configured to be enabled. This is a configuration value, not a user preference.
+     */
+    suspend fun isAutofillEnabledByConfiguration(): Boolean
```

Comment:
This API is very odd. It's internally checking for `isInternal` and that should be left to the caller.
I would think these APIs are way simpler than they're right now. Their composability should be left to the caller imo.

Reply:
Summarising offline chat here, in reply to all the existing comments.

the TL;DR of that class it that whether autofill can be enabled varies depending on a whole bunch of things, such as:

- internal tester
- global feature flag enabled
- subfeature flag enabled
- device capabilities
- if user has device auth set
- which autofill subfeature you're querying
- which part of the codebase you're calling from
- etc...

and it's not scalable to have each caller know about and query the right things. so the purpose of this class is to centralise and contain the complexity so that other callers don't need to consider all of the many nuanced cases.

For `isInternal` check, it's one of the many factors that determine autofill availability. if it lived outside the class (which is possible) you have to proliferate that `isInternal` check across the codebase. i think this is worse, because now every caller has to be aware of internal testing and check its state, in addition to checking other states, and combining the results. it's also optimising for deletion: when we remove the internal testing check it's only one place it is removed; the rest of the codebase is insulated from its existence or later removal.